### ISSUE TITLE
Improve handling of larger data layers

### ISSUE THREAD

**Issue by mvl22** [nptscot/nptscot.github.io#258]
Some of the data layers are quite large in data terms, even for vector tiles.

Accordingly it would be sensible to show a loading spinner, as currently the user can be left with confusion as to current state. This is obviously a fairly standard UI thing we currently lack.

Should be straightforward, as shown by this [pattern](https://gist.github.com/malwoodsantoro/5185fa4f463259ed285a9f4e00700690).

**Comment by Robinlovelace** [comment]
Really? I've never seen datasets take time to load. Will help if you provide a .gif/.mp4 or similar to show what you mean.

**Comment by mvl22** [comment]
I frequently observe the tiles for the default layer appearing one by one when zoomed out, even on a broadband connection.

It's worth trying the site on a non-JANET connection.

**Comment by Robinlovelace** [comment]
Are you suggesting that the spinner would stay on the screen until *all* the layer is shown? I think the fact that the linestrings appear incrementally, broadly from left to right, in your video gives a clear sense of "it's loading". This does not seem like a priority if worth doing at all. Let's check with the client before spending time on is my suggestion.

**Comment by mvl22** [comment]
I think part of the reason for this performance is the way that the PMTiles library HTTP range reading takes quite large ranges at a time - the first request I am seeing is a whopping 4.3MB - that is a lot of data to download and convert to a render pipeline:

![Screenshot 2024-10-28 at 09 41 23](https://github.com/user-attachments/assets/00abc94f-067e-45f4-a164-e379451443e8)

Looking at the documentation, there does not appear to be any way to control this to make better use of parallel browser requests:

https://docs.protomaps.com/pmtiles/maplibre

It may be that this is already as optimised as possible and we simply have a lot of data to shown that cannot be optimised.

I note that the approach of Tippecanoe for using file-per-tile rendering limits tile sizes to 500MB by default.

**Comment by Robinlovelace** [comment]
Great to have feature requests like this, and good to read about what's going on under the hood. Good question to ask in the meeting on Weds?

**Comment by mvl22** [comment]
> Are you suggesting that the spinner would stay on the screen until all the layer is shown?

No, that there would be a small spinner next to the layer box, straightforward stuff.

The video I showed was clipped to enable Github to load it  - it can easily be 20s before the whole site with that default layer is shown in the default view. That kind of time load normally requires an indicator for user assurance.

Incidentally, I just reproduced the loading problem with Edinburgh. It's simply very dense data. Sometimes it just isn't even showing at all when zoomed out. I think the browser is basically giving up.

**Comment by Robinlovelace** [comment]
> Incidentally, I just reproduced the loading problem with Edinburgh. It's simply very dense data. Sometimes it just isn't even showing at all when zoomed out. I think the browser is basically giving up.

That is very interesting. 

Worth posting a question and asking the author of PMtiles, @bdon (apologies for tagging) in a discussion?

More info, maybe we should reopen this if we find a way to tackle it?

https://github.com/nptscot/nptscot.github.io/issues/244

One more comment: I think having separate layers for fastest, balanced, quietest routes will help.

**Comment by Robinlovelace** [comment]
Link to PMTiles discussion (got really quick feedback from author here): https://github.com/protomaps/PMTiles/discussions/234

**Comment by bdon** [comment]
>4MB is really large for a single tile, you should optimize your tileset to have less properties or more simplified features.

Parallel browser requests require HTTP/2 which is not supported on Azure Blob storage. 

You could try running this proxy https://docs.protomaps.com/deploy/azure which will support HTTP/2 and eliminate the overhead of decoding PMTiles on the client.

**Comment by Robinlovelace** [comment]
@mvl22 can you test with the simplified network toggled to on? From my tests that resolves the issue:

![image](https://github.com/user-attachments/assets/9d722ce1-13f7-48f7-b784-4c07e3b470e5)

**Comment by mvl22** [comment]
> `--maximum-tile-bytes=5000000`

Is that deliberate?

5MB is a large size, and I'm not surprised this is giving rather slow performance. That is ten times larger than the Tippecanoe default, i.e. 500K.

(By parallelisation, I meant the way in which browsers can make multiple requests in parallel.)

**Comment by mvl22** [comment]
> @mvl22 can you test with the simplified network toggled to on? From my tests that resolves the issue:

Seems marginally better, but it's still a large data size, so defaulting to that doesn't solve the problem.

**Comment by Robinlovelace** [comment]
> `--maximum-tile-bytes=5000000`

That was inherited from @mem48. Let's try reducing it, e.g. to `2000000`, in the next full build and see how that goes.

Are you happy to give that a bash @wangzhao0217 ?

**Comment by mvl22** [comment]
I would try removing entirely to start with, and only add it back in, with a lower figure, if you find it is needed. The default is probably designed for various optimisations so best to use it unless the data won't fit.

**Comment by mem48** [comment]
A few useful details

👍 to @mvl22 's idea of a spinner, it was on my original to-do list and never got around to it. This loading problem is never going to go away as there will always be one more layer to add or someone on a slow connection.


 `--maximum-tile-bytes=5000000` is not the default, which as @mvl22 points out is `500000`. I jacked it up for development as the default looked terrible, and then didn't get around to reducing. Feel free to reduce, but you will absolutely notice it in Edinburgh and Glasgow. In other places, the tiles are already not hitting the limit, so a lower limit will not make a difference.

The issue with a lower limit is that Tippecanoe is not intelligent, so it will arbitrarily drop roads and produce confusing results. I have a fix for this that was implemented in the PBCC https://www.carbon.place/legacy/#8/51.482/-0.151. I made multiple rnets at lower zooms; only roads with high flow (under go Dutch) are included. This works well for urban areas with a mix of high/low flow roads, but in rural areas, you can have nothing until about Zoom 9. Robins pointed to a few ideas in the Tippecanoe docs, but my experience is that they are not great in practice. 

![image](https://github.com/user-attachments/assets/4d1ce114-3b77-4507-9c3c-35a27e36fe88)
This map is for England and Wales, but the low flow in Wales filters all roads out.

I'm sure you could come up with smarter rules. However, the point is that the filter needs to be applied before using Tippecanoe, as the options in Tippecanoe are too blunt. 


The OS Zoom Stack used for the basemap is also large-ish and could be further optimised. In particular, the buildings layer used for the zone data and the DataZones adds some overhead. But also the roads layers could be better filtered at low zoom levels.

More importantly, the pmtiles are in simple blob storage, and there is no CDN in front of them. This means there are none of the niceties of preloading, compression, or a proper cache policy that could easily offer significant performance benefits. At the moment, it is set to no-cache, which is terrible for any repeat user. 

You can also look at pre-fetching / pre-loading. Currently no map data can be downloaded until the maplibre and pmtiltes libraries have been downloaded and their large amount of javascript has been run, which takes a while. It may be possible to pre-fetch a few tiles of data which the libraries are loading, eliminating this render-blocking issue. For Carbon & Place one of the benefits of having a home page without a map is that the homepage loads quickly and then quietly downloads things that the maps need before the users has clicked on any map https://github.com/PlaceBasedCarbonCalculator/PlaceBasedCarbonCalculator.github.io/blob/f24cc31615c4dad4da65cf4523dba5f85700fc3a/index.html#L193

**Comment by mem48** [comment]
Oh and I think @mvl22 changed the welcome modal so that it doesn't show every time. That was absolutly put in to hide the map while the data was loading in. So you coudl disable that feature or move to a dedicated homepage + preloading

**Comment by mvl22** [comment]
> Oh and I think @mvl22 changed the welcome modal so that it doesn't show every time. That was absolutly put in to hide the map while the data was loading in. So you coudl disable that feature or move to a dedicated homepage + preloading

I don't think using a forced welcome message as a workaround is the way forward. A loading spinner is the least worst while the data loading is slow, as that does represent the state explicitly. Users may, for instance, choose to zoom in to avoid that once they are able to see there is a delay.

Thanks for the above info.

**Comment by Robinlovelace** [comment]
Useful info for sure. I suggest we do try the brute force approach of reducing the max tile size, if links are being dropped randomly that's not the end of the world.

**Comment by Robinlovelace** [comment]
> If the tiles come out too big, the --drop-densest-as-needed option will make Tippecanoe try dropping what should be the least visible features at each zoom level. (If it drops too many features, use -x to leave out some feature attributes that you didn't really need.)

And we have drop-densest so let's see what happens. `--drop-by-lowest-value` would be useful.

**Comment by mvl22** [comment]
How is this now?

I loaded the site freshly and Edinburgh appeared reliably, and what felt like quite a bit faster, unless I am imagining this.

**Comment by Robinlovelace** [comment]
Fine to close it by me.  I don't think that adding a spinner as suggested above is a priority given other issues like https://github.com/nptscot/nptscot.github.io/issues/264, https://github.com/nptscot/nptscot.github.io/issues/266 and https://github.com/nptscot/nptscot.github.io/issues/294

**Comment by Robinlovelace** [comment]
There quicker wins tracked in https://github.com/nptscot/npt/issues/561
